src/keymap.tsx

```diff
@@ -1,12 +1,65 @@
 import React from 'react';
-import CodeMirror from 'codemirror';
+import CodeMirror, { Editor } from 'codemirror';
 import SHARED from './shared';
 import {delete_, copy, paste, InsertTarget,
   ReplaceNodeTarget, OverwriteTarget, activateByNid} from './actions';
 import {partition, getRoot, skipCollapsed, say, mac,
   getLastVisibleNode, preambleUndoRedo, playSound, BEEP} from './utils';
 import {findAdjacentDropTargetId as getDTid} from './components/DropTarget';
 
+import type { AppDispatch } from './store';
+import type { AST, ASTNode } from './ast';
+import type { RootState } from './reducers';
+import type { EnhancedNodeProps } from './components/Node';
+import type { BlockEditorProps } from './ui/BlockEditor';
+
+/**
+ * This is completely bananas. This InputEnv type is what
+ * the keyDown() function expects for it's "env" parameter.
+ * 
+ * In theory, keyDown() is always passed the "this" value
+ * from inside of a Node react component, with a few extra
+ * things added on dynamically.
+ * 
+ * keyDown() then tacks on even more attributes to this "env"
+ * object so that all the handlers in commandMap can access
+ * these additional attributes.
+ * 
+ * This should be refactored to something that's is less
+ * abusive to the properties of the Node component.
+ */
+type InputEnv = {
+  // added by BlockEditor before calling keyDown()
+  showDialog: BlockEditorProps['showDialog'],
+  toolbarRef: BlockEditorProps['toolbarRef'],
+
+  // defined by Node react component
+  isLocked():boolean;
+  handleMakeEditable: (e?:Event) => void;
+  setRight():boolean;
+  setLeft():boolean;
+
+  // These somehow come from somewhere. Either a BlockEditor or a Node
+  // elements presumably.
+  cur: CodeMirror.Position;
+  props:{
+    dispatch: AppDispatch,
+    activateByNid: (nid: number, options?:{allowMove: boolean, record?: boolean})=>void,
+  } & EnhancedNodeProps,
+  node?: ASTNode,
+  dispatch: AppDispatch,
+  ast: AST;
+
+  // these get tacked on by the keyDown function
+  fastSkip?: (next: (node: ASTNode)=>ASTNode) => ASTNode,
+  activate?: (n: ASTNode|null|undefined, options:{allowMove: boolean, record: boolean}) => void,
+  activateNoRecord?: (node?: ASTNode) => void;
+}
+
+type Env = InputEnv & InputEnv['props'] & Pick<RootState, "ast"|"selections"> & {
+  state: RootState
```

Comment:
This is not great. I'd like to come up with a better way to do this. In theory, `env` is supposed to be the `this` value from inside a `<Node />` component. But then a bunch of stuff gets tacked onto that object, which means we are modifying `<Node />` objects at runtime, which makes the whole thing very fragile.

I'll have to think about this a bit to come up with something more bulletproof.

Reply:
I agree: it felt like a filthy hack when I was doing it! But I couldn't figure out a better way to re-use the same key event code across nodes and the top-level CM environment. We had like ~250 duplicate LOC, and it was a maintenance nightmare. I wonder if there's a way of treating CM itself as the "parent root node" of the whole tree, so that _every_ event comes from an Node component?

Reply:
I think the first step would be to specify the intended interface, which is probably more constrained than "whatever is available on a <node> component". If the <node> component implements that interface, then you can pass the node component in directly. If the `env` is coming from something else that doesn't implement that interface, then you use the adapter/facade pattern to create an object that does implement that interface and pass that object in instead.

The particularly fragile part IMO is the use of code like `env = Object.assign(env, env.props)`, which mutates `env` and flattens its properties. If `env` looks like
```js
{
  dispatch: ()=> { console.log("do something"); },
  props: {
    dispatch: ()=> { console.log("do something else"); },
  }
}
```
then not only is it very non-obvious which `dispatch()` function you are calling, you've also redefined the `dispatch()` function on the `<node>` while it is still in use. Right now these naming collisions don't happen (hopefully?) but they are extremely easy to accidentally introduce and then impossible to debug.